File: tests/fixtures/open_redirect/java/SafeRelativeRedirect.java
// Safe: request param routed through `ensureRelativeUrl` which enforces a
// leading `/` and rejects scheme-prefixed values (relative-only path).
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class SafeRelativeRedirect {
    public static String ensureRelativeUrl(String raw) {
        if (raw == null || !raw.startsWith("/") || raw.startsWith("//")) {
            return "/";
        }
        return raw;
    }

    public void handle(HttpServletRequest req, HttpServletResponse res) throws IOException {
        String target = req.getParameter("next");
        String safe = ensureRelativeUrl(target);
        res.sendRedirect(safe);
    }
}

File: tests/fixtures/open_redirect/rust/safe_relative_redirect.rs
// Safe: tainted value routed through `ensure_relative_url` which enforces
// a leading `/` and rejects scheme-prefixed or protocol-relative values
// (relative-only path).
use axum::response::Redirect;
use std::env;

fn ensure_relative_url(raw: &str) -> String {
    if !raw.starts_with('/') || raw.starts_with("//") {
        return "/".to_string();
    }
    raw.to_string()
}

fn bounce() -> Redirect {
    let next = env::var("NEXT").unwrap_or_default();
    let safe = ensure_relative_url(&next);
    Redirect::permanent(&safe)
}
